// popmotion/lib/animations/keyframes.js
import { interpolate } from "../utils/interpolate";
import { easeInOut } from "../easing";
export function defaultEasing(values, easing) {
    return values.map(function () { return easing || easeInOut; }).splice(0, values.length - 1);
}
export function defaultOffset(values) {
    var numValues = values.length;
    return values.map(function (_value, i) {
        return i !== 0 ? i / (numValues - 1) : 0;
    });
}
export function convertOffsetToTimes(offset, duration) {
    return offset.map(function (o) { return o * duration; });
}
var KeyframesAnimator = (function () {
    function KeyframesAnimator(options) {
        this.isComplete = false;
        this.updateOptions(options);
        var _a = this.options, from = _a.from, to = _a.to;
        this.values = Array.isArray(to) ? to : [from, to];
        this.createInterpolator();
    }
    KeyframesAnimator.prototype.createInterpolator = function () {
        var _a = this.options, duration = _a.duration, ease = _a.ease, offset = _a.offset;
        ease = Array.isArray(ease) ? ease : defaultEasing(this.values, ease);
        offset = offset || defaultOffset(this.values);
        var times = convertOffsetToTimes(offset, duration);
        this.interpolator = interpolate(times, this.values, { ease: ease });
    };
    KeyframesAnimator.prototype.update = function (t) {
        var duration = this.options.duration;
        this.isComplete = t >= duration;
        return this.interpolator(t);
    };
    KeyframesAnimator.prototype.updateOptions = function (_a) {
        var _b = _a.from, from = _b === void 0 ? 0 : _b, _c = _a.to, to = _c === void 0 ? 1 : _c, ease = _a.ease, offset = _a.offset, _d = _a.duration, duration = _d === void 0 ? 300 : _d;
        this.options = { from: from, to: to, ease: ease, offset: offset, duration: duration };
    };
    KeyframesAnimator.prototype.flipTarget = function () {
        this.values.reverse();
        this.createInterpolator();
    };
    KeyframesAnimator.needsInterpolation = function () { return false; };
    KeyframesAnimator.uniqueOptionKeys = new Set([
        "duration",
        "ease",
    ]);
    return KeyframesAnimator;
}());
export { KeyframesAnimator };
